In [1]:
import pandas as pd

In [2]:
train = pd.read_csv('./datasets/mother.csv')
train.head()

Unnamed: 0  Overall Qual  Year Built  Year Remod/Add  Mas Vnr Area  \
0           0     -0.078644    0.142227        0.989479      1.092329   
1           1      0.622656    0.805126        0.609090      0.191491   
2           2     -0.779944   -0.620106        1.084576     -0.565901   
3           3     -0.779944    1.136575        1.084576     -0.565901   
4           4     -0.078644   -2.376787        0.418896     -0.565901   

   Total Bsmt SF  1st Flr SF  Gr Liv Area  Full Bath  TotRms AbvGrd  \
0      -0.741232   -1.108838    -0.040634   0.769779      -0.279441   
1      -0.322705   -0.634510     1.244529   0.769779       1.002738   
2      -0.002130   -0.271195    -0.884084  -1.051232      -0.920531   
3      -1.500370   -1.060900    -0.110588   0.769779       0.361648   
4      -0.850317   -0.841397    -0.108589   0.769779      -0.279441   

     ...      Garage Area Yr Sold  Garage Area baths  Yr Sold homeage  \
0    ...                 0.008581           0.496281        -0.067210   
1    ...                 0.396922           0.848706        -0.763445   
2    ...                -1.053629          -1.083758         0.695100   
3    ...                -0.339305           0.181616        -1.061526   
4    ...                 0.050327           0.127913         2.451726   

   Yr Sold baths  Yr Sold garageage   baths^2  Foundation_PConc  Bsmt Qual_Ex  \
0       1.160521           0.203566  1.138397                 0             0   
1       1.158568          -0.686130  1.138397                 1             0   
2      -1.194594           1.133449 -1.046041                 0             0   
3       1.160521          -1.049756  1.138397                 1             0   
4       0.375483           0.971731  0.202209                 1             0   

   Kitchen Qual_Ex  SalePrice  
0                0     130500  
1                0     220000  
2                0     109000  
3                0     174000  
4                0     138500  

[5 rows x 169 columns]

In [3]:
from sklearn import metrics
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LassoCV, RidgeCV, ElasticNetCV
from sklearn.preprocessing import StandardScaler, PowerTransformer
from sklearn.metrics import r2_score

In [5]:
len(train.columns)

169

In [28]:
features = ['Overall Qual',
 'Year Built',
 'Year Remod/Add',
 'Mas Vnr Area',
 'Total Bsmt SF',
 '1st Flr SF',
 'Gr Liv Area',
 'Full Bath',
 'TotRms AbvGrd',
 'Garage Yr Blt',
 'Garage Cars',
 'Garage Area',
 'homeage',
 'baths',
 'garageage',
 'Lot Frontage Overall Qual',
 'Lot Frontage Mas Vnr Area',
 'Lot Frontage Gr Liv Area',
 'Lot Frontage Full Bath',
 'Lot Frontage Garage Cars',
 'Lot Frontage Garage Area',
 'Lot Frontage baths',
 'Lot Area Overall Qual',
 'Lot Area Garage Cars',
 'Overall Qual^2',
 'Overall Qual Overall Cond',
 'Overall Qual Year Built',
 'Overall Qual Year Remod/Add',
 'Overall Qual Mas Vnr Area',
 'Overall Qual BsmtFin SF 1',
 'Overall Qual Total Bsmt SF',
 'Overall Qual 1st Flr SF',
 'Overall Qual Gr Liv Area',
 'Overall Qual Full Bath',
 'Overall Qual Bedroom AbvGr',
 'Overall Qual Kitchen AbvGr',
 'Overall Qual TotRms AbvGrd',
 'Overall Qual Fireplaces',
 'Overall Qual Garage Yr Blt',
 'Overall Qual Garage Cars',
 'Overall Qual Garage Area',
 'Overall Qual Yr Sold',
 'Overall Qual baths',
 'Overall Cond Total Bsmt SF',
 'Overall Cond 1st Flr SF',
 'Overall Cond Gr Liv Area',
 'Overall Cond Garage Cars',
 'Overall Cond Garage Area',
 'Overall Cond baths',
 'Year Built^2',
 'Year Built Year Remod/Add',
 'Year Built Mas Vnr Area',
 'Year Built Total Bsmt SF',
 'Year Built 1st Flr SF',
 'Year Built Gr Liv Area',
 'Year Built Full Bath',
 'Year Built TotRms AbvGrd',
 'Year Built Garage Yr Blt',
 'Year Built Garage Cars',
 'Year Built Garage Area',
 'Year Built Yr Sold',
 'Year Built homeage',
 'Year Built baths',
 'Year Built garageage',
 'Year Remod/Add^2',
 'Year Remod/Add Mas Vnr Area',
 'Year Remod/Add Total Bsmt SF',
 'Year Remod/Add 1st Flr SF',
 'Year Remod/Add Gr Liv Area',
 'Year Remod/Add Full Bath',
 'Year Remod/Add TotRms AbvGrd',
 'Year Remod/Add Garage Yr Blt',
 'Year Remod/Add Garage Cars',
 'Year Remod/Add Garage Area',
 'Year Remod/Add Yr Sold',
 'Year Remod/Add homeage',
 'Year Remod/Add baths',
 'Year Remod/Add garageage',
 'Mas Vnr Area Total Bsmt SF',
 'Mas Vnr Area 1st Flr SF',
 'Mas Vnr Area Gr Liv Area',
 'Mas Vnr Area Full Bath',
 'Mas Vnr Area Bedroom AbvGr',
 'Mas Vnr Area TotRms AbvGrd',
 'Mas Vnr Area Fireplaces',
 'Mas Vnr Area Garage Yr Blt',
 'Mas Vnr Area Garage Cars',
 'Mas Vnr Area Garage Area',
 'Mas Vnr Area Yr Sold',
 'Mas Vnr Area baths',
 'BsmtFin SF 1 Full Bath',
 'BsmtFin SF 1 Garage Cars',
 'BsmtFin SF 1 baths',
 'Total Bsmt SF 1st Flr SF',
 'Total Bsmt SF Gr Liv Area',
 'Total Bsmt SF Full Bath',
 'Total Bsmt SF Bedroom AbvGr',
 'Total Bsmt SF TotRms AbvGrd',
 'Total Bsmt SF Fireplaces',
 'Total Bsmt SF Garage Yr Blt',
 'Total Bsmt SF Garage Cars',
 'Total Bsmt SF Garage Area',
 'Total Bsmt SF Yr Sold',
 'Total Bsmt SF baths',
 '1st Flr SF^2',
 '1st Flr SF Gr Liv Area',
 '1st Flr SF Full Bath',
 '1st Flr SF TotRms AbvGrd',
 '1st Flr SF Fireplaces',
 '1st Flr SF Garage Yr Blt',
 '1st Flr SF Garage Cars',
 '1st Flr SF Garage Area',
 '1st Flr SF Yr Sold',
 '1st Flr SF baths',
 'Gr Liv Area^2',
 'Gr Liv Area Full Bath',
 'Gr Liv Area TotRms AbvGrd',
 'Gr Liv Area Fireplaces',
 'Gr Liv Area Garage Yr Blt',
 'Gr Liv Area Garage Cars',
 'Gr Liv Area Garage Area',
 'Gr Liv Area Yr Sold',
 'Gr Liv Area baths',
 'Full Bath^2',
 'Full Bath TotRms AbvGrd',
 'Full Bath Fireplaces',
 'Full Bath Garage Yr Blt',
 'Full Bath Garage Cars',
 'Full Bath Garage Area',
 'Full Bath Yr Sold',
 'Full Bath baths',
 'Bedroom AbvGr Garage Cars',
 'Bedroom AbvGr Garage Area',
 'Kitchen AbvGr Garage Cars',
 'Kitchen AbvGr Garage Area',
 'Kitchen AbvGr homeage',
 'Kitchen AbvGr garageage',
 'TotRms AbvGrd Fireplaces',
 'TotRms AbvGrd Garage Yr Blt',
 'TotRms AbvGrd Garage Cars',
 'TotRms AbvGrd Garage Area',
 'TotRms AbvGrd Yr Sold',
 'TotRms AbvGrd baths',
 'Fireplaces Garage Cars',
 'Fireplaces Garage Area',
 'Fireplaces baths',
 'Garage Yr Blt^2',
 'Garage Yr Blt Garage Cars',
 'Garage Yr Blt Garage Area',
 'Garage Yr Blt Yr Sold',
 'Garage Yr Blt homeage',
 'Garage Yr Blt baths',
 'Garage Yr Blt garageage',
 'Garage Cars^2',
 'Garage Cars Garage Area',
 'Garage Cars Yr Sold',
 'Garage Cars baths',
 'Garage Area^2',
 'Garage Area Yr Sold',
 'Garage Area baths',
 'Yr Sold homeage',
 'Yr Sold baths',
 'Yr Sold garageage',
 'baths^2',
'Foundation_PConc',
'Bsmt Qual_Ex', 
'Kitchen Qual_Ex']

In [29]:
len(features)

167

In [30]:
X = train[features]
y = train['SalePrice']

In [31]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [32]:
#instantiate the models
lr = LinearRegression()
lasso = LassoCV()
ridge = RidgeCV()
elastic = ElasticNetCV()

# The four models- Cross Val Score

In [33]:
cross_val_score(lr, X_train, y_train, cv= 10).mean()

0.862850248245848

In [34]:
cross_val_score(lasso, X_train, y_train, cv= 10).mean()

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify 

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Obje

0.7825381459907108

In [35]:
cross_val_score(ridge, X_train, y_train, cv= 10).mean()

0.8666884410407742

In [36]:
cross_val_score(elastic, X_train, y_train, cv= 10).mean()

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a va

0.6003718329968822

# Proceed with Ridge 

In [37]:
ridge.fit(X_train, y_train)

RidgeCV(alphas=array([ 0.1,  1. , 10. ]), cv=None, fit_intercept=True,
    gcv_mode=None, normalize=False, scoring=None, store_cv_values=False)

In [38]:
#cv scores
cross_val_score(ridge, X_train, y_train, cv=10).mean()

0.8666884410407742

In [39]:
#train score
ridge.score(X_train, y_train)

0.9283344627138852

In [40]:
#test score
ridge.score(X_test, y_test)

0.9031217146246528

In [41]:
import numpy as np

In [42]:
#train predictions 
predictions = ridge.predict(X_train)

In [43]:
metrics.mean_squared_error(y_train, predictions)

459117493.7416403

In [44]:
import numpy as np
np.sqrt(metrics.mean_squared_error(y_train, predictions))

21427.027179280853

In [45]:
#test prdictions 
predictions = ridge.predict(X_test)
metrics.mean_squared_error(y_test, predictions)

579082877.2793851

In [46]:
np.sqrt(metrics.mean_squared_error(y_test, predictions))

24064.140900505572

# Evaluating Metrics using Ridge Regression

- Train score: .9283
- Test Score: .9031
- RMSE Train: 21,427
- RMSE Test: 24,064